# Tucil 1 IF3270 Machine Learning dengan Python

### 13519153 Maximillian Lukman - 13519162 Irvin Andryan Pratomo

In [98]:
from sklearn import datasets, tree, model_selection, preprocessing, metrics, svm, cluster, linear_model, neural_network

In [99]:
import six
import sys
sys.modules['sklearn.externals.six'] = six

In [100]:
from id3 import Id3Estimator, export_text

In [101]:
import pandas as pd

In [102]:
#load data from sklearn.datasets
cancer = datasets.load_breast_cancer()

In [103]:
#load data from csv file
tennis = pd.read_csv('play_tennis.csv')

In [104]:
#transform label for tennis
le = preprocessing.LabelEncoder()

tennis = tennis.drop('day', axis = 1)
tennis['outlook'] = le.fit_transform(tennis['outlook'])
tennis['temp'] = le.fit_transform(tennis['temp'])
tennis['humidity'] = le.fit_transform(tennis['humidity'])
tennis['wind'] = le.fit_transform(tennis['wind'])
tennis['play'] = le.fit_transform(tennis['play'])

In [105]:
X_cancer = cancer.data
y_cancer = cancer.target

In [106]:
X_tennis = tennis.drop('play', axis = 1)
y_tennis = tennis.play

In [107]:
#Set training data to 80% and test data to 20% for cancer
X_train_cancer, X_test_cancer, y_train_cancer, y_test_cancer = model_selection.train_test_split(X_cancer,y_cancer,test_size=0.2,shuffle=True)

In [108]:
#Set training data to 80% and test data to 20% for tennis
X_train_tennis, X_test_tennis, y_train_tennis, y_test_tennis = model_selection.train_test_split(X_tennis,y_tennis,test_size=0.2,shuffle=True)

## Decision Tree Classifier

In [109]:
decision_tree = tree.DecisionTreeClassifier(criterion='entropy', max_depth=20)

In [110]:
#decision tree for tennis
decision_tree_tennis = decision_tree.fit(X_train_tennis, y_train_tennis)
r_tennis = tree.export_text(decision_tree_tennis, feature_names=tennis.columns[:-1].tolist())
print("Play Tennis")
print(r_tennis)

Play Tennis
|--- wind <= 0.50
|   |--- outlook <= 0.50
|   |   |--- class: 1
|   |--- outlook >  0.50
|   |   |--- outlook <= 1.50
|   |   |   |--- class: 0
|   |   |--- outlook >  1.50
|   |   |   |--- humidity <= 0.50
|   |   |   |   |--- class: 0
|   |   |   |--- humidity >  0.50
|   |   |   |   |--- class: 1
|--- wind >  0.50
|   |--- class: 1



In [111]:
prediction_tennis_decisionTreeClassifier = decision_tree_tennis.predict(X_test_tennis)
print("Decision Tree Classifier: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_decisionTreeClassifier))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_decisionTreeClassifier, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_decisionTreeClassifier, average="weighted"))

Decision Tree Classifier: Tennis
Accuracy: 0.3333333333333333
F1 Macro avg: 0.25
F1 Weighted avg: 0.16666666666666666


In [112]:
#decision tree for cancer
decision_tree_cancer = decision_tree.fit(X_train_cancer, y_train_cancer)
r_cancer = tree.export_text(decision_tree_cancer, feature_names=cancer.feature_names.tolist())
print("Breast Cancer")
print(r_cancer)

Breast Cancer
|--- worst radius <= 16.80
|   |--- worst concave points <= 0.14
|   |   |--- mean texture <= 21.43
|   |   |   |--- class: 1
|   |   |--- mean texture >  21.43
|   |   |   |--- worst area <= 643.25
|   |   |   |   |--- class: 1
|   |   |   |--- worst area >  643.25
|   |   |   |   |--- concave points error <= 0.02
|   |   |   |   |   |--- compactness error <= 0.01
|   |   |   |   |   |   |--- mean perimeter <= 90.44
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- mean perimeter >  90.44
|   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- compactness error >  0.01
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |--- concave points error >  0.02
|   |   |   |   |   |--- class: 0
|   |--- worst concave points >  0.14
|   |   |--- worst texture <= 25.67
|   |   |   |--- concave points error <= 0.01
|   |   |   |   |--- area error <= 20.88
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- area error >  20.88
|   |   |   |   |  

In [113]:
prediction_cancer_decisionTreeClassifier = decision_tree_cancer.predict(X_test_cancer)
print("Decision Tree Classifier: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_decisionTreeClassifier))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_decisionTreeClassifier, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_decisionTreeClassifier, average="weighted"))

Decision Tree Classifier: Cancer
Accuracy: 0.9385964912280702
F1 Macro avg: 0.9246814535158093
F1 Weighted avg: 0.9377445501436461


## Id3Estimator

In [114]:
estimator = Id3Estimator()

In [115]:
estimator_tennis = estimator.fit(X_train_tennis, y_train_tennis)

In [116]:
prediction_tennis_Id3 = estimator_tennis.predict(X_test_tennis)
print("Id3Estimator: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_Id3))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_Id3, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_Id3, average="weighted"))

Id3Estimator: Tennis
Accuracy: 0.3333333333333333
F1 Macro avg: 0.25
F1 Weighted avg: 0.16666666666666666


In [117]:
estimator_tennis_tree = export_text(estimator_tennis.tree_, feature_names=tennis.columns[:-1].tolist())
print("Play Tennis")
print(estimator_tennis_tree)

Play Tennis

wind <=0.50
|   outlook <=0.50: 1 (2) 
|   outlook >0.50
|   |   temp <=1.50: 0 (2) 
|   |   temp >1.50
|   |   |   humidity <=0.50: 0 (1) 
|   |   |   humidity >0.50: 1 (1) 
wind >0.50: 1 (5) 



In [118]:
estimator_cancer = estimator.fit(X_train_cancer, y_train_cancer)

In [119]:
estimator_cancer_tree = export_text(estimator_cancer.tree_, cancer['feature_names'])
print("Breast Cancer")
print(estimator_cancer_tree)

Breast Cancer

worst radius <=16.80
|   worst concave points <=0.14
|   |   mean texture <=21.43: 1 (215) 
|   |   mean texture >21.43
|   |   |   worst area <=643.25: 1 (30) 
|   |   |   worst area >643.25
|   |   |   |   mean smoothness <=0.09
|   |   |   |   |   mean radius <=13.45
|   |   |   |   |   |   mean perimeter <=86.26: 0 (2) 
|   |   |   |   |   |   mean perimeter >86.26: 1 (1) 
|   |   |   |   |   mean radius >13.45: 1 (10) 
|   |   |   |   mean smoothness >0.09: 0 (2) 
|   worst concave points >0.14
|   |   worst texture <=25.67
|   |   |   concave points error <=0.01
|   |   |   |   mean texture <=16.98: 0 (2) 
|   |   |   |   mean texture >16.98: 1 (2) 
|   |   |   concave points error >0.01: 1 (10) 
|   |   worst texture >25.67
|   |   |   mean concavity <=0.09
|   |   |   |   mean radius <=13.34: 0 (1) 
|   |   |   |   mean radius >13.34: 1 (2) 
|   |   |   mean concavity >0.09: 0 (22) 
worst radius >16.80
|   mean texture <=14.99
|   |   mean compactness <=0.13: 1 (

In [120]:
prediction_cancer_Id3 = estimator_cancer.predict(X_test_cancer)
print("Id3Estimator: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_Id3))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_Id3, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_Id3, average="weighted"))

Id3Estimator: Cancer
Accuracy: 0.9122807017543859
F1 Macro avg: 0.8952205882352942
F1 Weighted avg: 0.9122807017543859


## K-Means

In [121]:
kmeans = cluster.KMeans(n_clusters=2)

In [122]:
kmeans_tennis = kmeans.fit(X_train_tennis,y_train_tennis)

In [123]:
kmeans_tennis_centroid = pd.DataFrame(kmeans_tennis.cluster_centers_.transpose())
kmeans_tennis_centroid.index = tennis.columns[:-1].tolist()
kmeans_tennis_centroid.columns = ["Centroid 1", "Centroid 2"]
print(kmeans_tennis_centroid)

          Centroid 1  Centroid 2
outlook          1.2    0.666667
temp             1.8    0.333333
humidity         0.2    0.833333
wind             0.2    0.666667


In [124]:
prediction_tennis_kmeans = kmeans_tennis.predict(X_test_tennis)
print("K-Means: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_kmeans))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_kmeans, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_kmeans, average="weighted"))

K-Means: Tennis
Accuracy: 0.6666666666666666
F1 Macro avg: 0.4
F1 Weighted avg: 0.5333333333333333


In [125]:
kmeans_cancer = kmeans.fit(X_train_cancer,y_train_cancer)

In [126]:
kmeans_cancer_centroid = pd.DataFrame(kmeans_cancer.cluster_centers_.transpose())
kmeans_cancer_centroid.index = cancer['feature_names'].tolist()
kmeans_cancer_centroid.columns = ["Centroid 1", "Centroid 2"]
print(kmeans_cancer_centroid)

                         Centroid 1   Centroid 2
mean radius               12.630837    19.587429
mean texture              18.618429    21.724476
mean perimeter            81.683886   129.484762
mean area                501.985429  1206.572381
mean smoothness            0.094825     0.101158
mean compactness           0.092810     0.146516
mean concavity             0.064946     0.175317
mean concave points        0.034320     0.101167
mean symmetry              0.178605     0.190745
mean fractal dimension     0.063443     0.060226
radius error               0.302593     0.725070
texture error              1.196623     1.239931
perimeter error            2.166278     5.121648
area error                24.009317    92.783714
smoothness error           0.007092     0.006614
compactness error          0.023842     0.031844
concavity error            0.029107     0.042084
concave points error       0.010629     0.015767
symmetry error             0.020467     0.020688
fractal dimension er

In [127]:
prediction_cancer_kmeans = kmeans_cancer.predict(X_test_cancer)
print("K-Means: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_kmeans))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_kmeans, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_kmeans, average="weighted"))

K-Means: Cancer
Accuracy: 0.10526315789473684
F1 Macro avg: 0.09832506203473944
F1 Weighted avg: 0.06640982107875146


## Logistic Regression

In [128]:
lr = linear_model.LogisticRegression(class_weight='balanced', max_iter=2500, random_state=0)

In [129]:
lr_tennis = lr.fit(X_train_tennis,y_train_tennis)

In [130]:
prediction_tennis_lr = lr_tennis.predict(X_test_tennis)
print("Logistic Regression: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_lr))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_lr, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_lr, average="weighted"))

Logistic Regression: Tennis
Accuracy: 0.6666666666666666
F1 Macro avg: 0.6666666666666666
F1 Weighted avg: 0.6666666666666666


In [131]:
lr_tennis_coefficient = pd.DataFrame(lr_tennis.coef_.transpose())
lr_tennis_coefficient.index = tennis.columns[:-1].tolist()
lr_tennis_coefficient.columns = ["Coefficient"]
print("Play Tennis")
print("")
print(lr_tennis_coefficient)

Play Tennis

          Coefficient
outlook     -0.594406
temp         0.197587
humidity     0.511366
wind         1.033502


In [132]:
lr_cancer = lr.fit(X_train_cancer,y_train_cancer)

In [133]:
prediction_cancer_lr = lr_cancer.predict(X_test_cancer)
print("Logistic Regression: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_lr))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_lr, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_lr, average="weighted"))

Logistic Regression: Cancer
Accuracy: 0.956140350877193
F1 Macro avg: 0.9480448455017774
F1 Weighted avg: 0.9563202509966466


In [134]:
lr_cancer_coefficient = pd.DataFrame(lr_cancer.coef_.transpose())
lr_cancer_coefficient.index = cancer.feature_names.tolist()
lr_cancer_coefficient.columns = ["Coefficient"]
print("Breast Cancer")
print("")
print(lr_cancer_coefficient)

Breast Cancer

                         Coefficient
mean radius                 0.956030
mean texture                0.219475
mean perimeter             -0.396222
mean area                   0.030067
mean smoothness            -0.159605
mean compactness           -0.238834
mean concavity             -0.453349
mean concave points        -0.250831
mean symmetry              -0.272578
mean fractal dimension     -0.035377
radius error               -0.018472
texture error               1.699444
perimeter error             0.163954
area error                 -0.115496
smoothness error           -0.022168
compactness error           0.057191
concavity error            -0.028179
concave points error       -0.031378
symmetry error             -0.044235
fractal dimension error     0.013249
worst radius                0.228848
worst texture              -0.516900
worst perimeter            -0.084322
worst area                 -0.015815
worst smoothness           -0.321128
worst compactness      

## Neural Network

In [135]:
neural = neural_network.MLPClassifier(max_iter=1000)

In [136]:
neural_tennis = neural.fit(X_train_tennis,y_train_tennis)

In [137]:
prediction_tennis_neural = neural_tennis.predict(X_test_tennis)
print("Neural Network: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_neural))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_neural, average="macro"))
print("F1 Weighted:", metrics.f1_score(y_test_tennis, prediction_tennis_neural, average="weighted"))

Neural Network: Tennis
Accuracy: 0.3333333333333333
F1 Macro avg: 0.25
F1 Weighted: 0.16666666666666666


In [138]:
neural_tennis_coefficient = neural_tennis.coefs_
print("Play Tennis Coefficient")
print("")
print(neural_tennis_coefficient)

Play Tennis Coefficient

[array([[ 2.97168634e-01, -2.75355044e-01, -3.72041281e-01,
        -9.06265515e-02, -2.21376682e-01, -1.20596133e-01,
         1.88444635e-01, -3.54340810e-01,  6.61429659e-15,
        -5.28094546e-02, -3.81503808e-01,  4.49444827e-01,
        -1.95213540e-02,  1.37185919e-01,  1.20672159e-12,
         5.86934673e-17, -1.02676464e-01,  2.66385365e-01,
        -4.51563274e-01,  2.10394437e-01, -1.80523375e-01,
        -3.11013355e-01, -4.47417100e-01, -2.03966360e-01,
        -2.81119334e-01,  3.11124510e-01,  3.41082278e-01,
        -1.44133846e-01,  1.57995447e-01,  2.80851474e-01,
        -7.71803262e-15, -2.49291742e-01,  2.74176017e-01,
         5.13885296e-01,  2.69944061e-01, -1.26441045e-02,
         3.11645708e-01, -5.18790618e-01,  1.19110390e-01,
         1.57711164e-01,  3.39626225e-01, -7.42221896e-03,
        -2.71517511e-01, -1.34756345e-01, -3.38879919e-01,
        -9.10015162e-02, -5.04658360e-02,  2.71309403e-06,
         5.53541278e-02,  4.32

In [139]:
neural_cancer = neural.fit(X_train_cancer,y_train_cancer)

In [140]:
prediction_cancer_neural = neural_cancer.predict(X_test_cancer)
print("Neural Network: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_neural))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_neural, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_neural, average="weighted"))

Neural Network: Cancer
Accuracy: 0.956140350877193
F1 Macro avg: 0.9480448455017774
F1 Weighted avg: 0.9563202509966466


In [141]:
neural_cancer_coefficient = neural_cancer.coefs_
print("Breast Cancer Coefficient")
print("")
print(neural_tennis_coefficient)

Breast Cancer Coefficient

[array([[ 2.97168634e-01, -2.75355044e-01, -3.72041281e-01,
        -9.06265515e-02, -2.21376682e-01, -1.20596133e-01,
         1.88444635e-01, -3.54340810e-01,  6.61429659e-15,
        -5.28094546e-02, -3.81503808e-01,  4.49444827e-01,
        -1.95213540e-02,  1.37185919e-01,  1.20672159e-12,
         5.86934673e-17, -1.02676464e-01,  2.66385365e-01,
        -4.51563274e-01,  2.10394437e-01, -1.80523375e-01,
        -3.11013355e-01, -4.47417100e-01, -2.03966360e-01,
        -2.81119334e-01,  3.11124510e-01,  3.41082278e-01,
        -1.44133846e-01,  1.57995447e-01,  2.80851474e-01,
        -7.71803262e-15, -2.49291742e-01,  2.74176017e-01,
         5.13885296e-01,  2.69944061e-01, -1.26441045e-02,
         3.11645708e-01, -5.18790618e-01,  1.19110390e-01,
         1.57711164e-01,  3.39626225e-01, -7.42221896e-03,
        -2.71517511e-01, -1.34756345e-01, -3.38879919e-01,
        -9.10015162e-02, -5.04658360e-02,  2.71309403e-06,
         5.53541278e-02,  4.

## SVM

In [142]:
svc = svm.SVC()

In [143]:
svc_tennis = svc.fit(X_train_tennis,y_train_tennis)

In [144]:
prediction_tennis_svm = svc_tennis.predict(X_test_tennis)
print("SVM: Tennis")
print("Accuracy:", metrics.accuracy_score(y_test_tennis, prediction_tennis_svm))
print("F1 Macro avg:", metrics.f1_score(y_test_tennis, prediction_tennis_svm, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_tennis, prediction_tennis_svm, average="weighted"))

SVM: Tennis
Accuracy: 0.3333333333333333
F1 Macro avg: 0.25
F1 Weighted avg: 0.16666666666666666


In [145]:
svc_cancer = svc.fit(X_train_cancer,y_train_cancer)

In [146]:
prediction_cancer_svm = svc_cancer.predict(X_test_cancer)
print("SVM: Cancer")
print("Accuracy:", metrics.accuracy_score(y_test_cancer, prediction_cancer_svm))
print("F1 Macro avg:", metrics.f1_score(y_test_cancer, prediction_cancer_svm, average="macro"))
print("F1 Weighted avg:", metrics.f1_score(y_test_cancer, prediction_cancer_svm, average="macro"))

SVM: Cancer
Accuracy: 0.9385964912280702
F1 Macro avg: 0.9260220635950681
F1 Weighted avg: 0.9260220635950681


## Analisis 

### Perbandingan Nilai Akurasi dan F1 Score dari masing-masing Algoritma pada dataset Play Tennis

Setelah melakukan pembelajaran dengan beberapa algoritma di atas, nilai tertinggi untuk Accuracy diperoleh dengan menggunakan algoritma K-Means dan LogisticRegression yaitu sebesar 0.66. Untuk untuk F1 Score, nilai tertinggi diperoleh dengan menggunakan algoritma Logistic Regression sebesar 0.66 untuk macro average F1 dan weighted average F1.
Sedangkan nilai terendah untuk dataset Play Tennis didapat dengan keempat algoritma lainnya yaitu Decision Tree, Neural Network, ID3, dan SVM dengan skor 0.333 untuk Accuracy dan 0.25 / 0.167 untuk F1 Macro / Weighted average.

### Perbandingan Nilai Akurasi dan F1 Score dari masing-masing Algoritma pada dataset Breast Cancer

Setelah melakukan pembelajaran dengan beberapa algoritma di atas, didapatkan nilai tertinggi dengan menggunakan Neural Network dan LogisticRegression dengan 0.956 untuk Accuracy, 0.948 untuk F1 Macro avg, dan 0.956 untuk F1 Weighted avg. Sedangkan yang terendah didapat dengan pembelajaran KMeans dengan 0.105 untuk Accuracy, 0.098 untuk F1 Macro avg, dan 0.066 untuk F1 Weighted avg.

### Kesimpulan

Dengan melihat hasil atau nilai yang dihasilkan dari pembelajaran berbagai algoritma, kami melihat pola di mana skor yang dihasilkan untuk dataset breast cancer relatif lebih besar daripada untuk dataset play tennis. Hal ini kami asumsikan dikarenakan jumlah data yang sangat jauh berbeda di mana dataset breast cancer memiliki jumlah data yang banyak sedangkan data play tennis memiliki jumlah data yang sedikit sehingga proses pembelajaran untuk data play tennis kurang maksimal.